adapters/webhdfs-adapter/src/test/java/com/connexta/replication/adapters/webhdfs/WebHdfsClientTest.java

```diff
@@ -0,0 +1,258 @@
+
+public class WebHdfsClientTest {
+  private static final String HDFS_PORT = "12341";
+  private static final String BASE_URL = "http://localhost:" + HDFS_PORT;
+  private static final Logger LOGGER = LoggerFactory.getLogger(WebHdfsClientTest.class);
+  private static WebHdfsNodeAdapterFactory adapterFactory = new WebHdfsNodeAdapterFactory();
+  private static HdfsLocalCluster hdfsLocalCluster;
+  private static WebHdfsNodeAdapter adapter;
+
+  @BeforeClass
+  public static void setUpClass() throws Exception {
+    LOGGER.info("HDFS instance starting up!");
+    adapter = (WebHdfsNodeAdapter) adapterFactory.create(new URL(BASE_URL));
+    hdfsLocalCluster =
+        new HdfsLocalCluster.Builder()
+            .setHdfsNamenodePort(12345)
+            .setHdfsNamenodeHttpPort(Integer.parseInt(HDFS_PORT))
+            .setHdfsTempDir("embedded_hdfs")
+            .setHdfsNumDatanodes(1)
+            .setHdfsEnablePermissions(false)
+            .setHdfsFormat(true)
+            .setHdfsEnableRunningUserAsProxyUser(true)
+            .setHdfsConfig(new Configuration())
+            .build();
+    hdfsLocalCluster.start();
+  }
+
+  @AfterClass
+  public static void tearDownClass() throws Exception {
+    LOGGER.info("HDFS instance shutting down!");
+    adapter.close();
+    hdfsLocalCluster.stop();
+  }
+
+  @Test(expected = ReplicationException.class)
+  public void testInvalidUrlIsNotAvailable() throws MalformedURLException {
+    adapter = (WebHdfsNodeAdapter) adapterFactory.create(new URL("http://foobar:9999"));
+    adapter.isAvailable();
+  }
+
+  @Test(expected = ReplicationException.class)
+  public void testStoppedClusterIsNotAvailable() throws Exception {
+    hdfsLocalCluster.stop();
+    adapter.isAvailable();
+    // Not necessary to restart the cluster here because it will automatically be started before the
+    // next test. Note: this is not expected behavior given the @BeforeClass annotation on
+    // setUpClass().
+  }
+
+  @Test
+  public void testResourceFileIsCreated() throws Exception {
+    String testId = "123456789";
+    String testName = "testresource";
+    String filename = String.format("%s_%s.txt", testId, "946684800000");
+    CreateStorageRequest createStorageRequest = generateTestStorageRequest(testId, testName);
+
+    adapter.createResource(createStorageRequest);
+    verifyFileExists(filename);
+  }
+
+  @Test
+  public void testInvalidHostnameFails() throws IOException {
+    String testId = "123456789";
+    String testName = "testresource";
+    CreateStorageRequest createStorageRequest = generateTestStorageRequest(testId, testName);
+
+    WebHdfsNodeAdapter badAdapter =
+        (WebHdfsNodeAdapter) adapterFactory.create(new URL("http://foobar:12341"));
+    boolean createResourceSuccessful = badAdapter.createResource(createStorageRequest);
+    assertThat(createResourceSuccessful, is(false));
+    badAdapter.close();
+    // TODO - figure out a good way to track error messages in the logger
```

Comment:
Can't we just use `LOGGER.error` here?

Reply:
Ahh maybe that could've been worded better. I meant that currently there's no easy way to tell when a specific exception gets thrown because it ends up getting caught in a top-level method and a message gets printed to the logger instead. So the only way I could think of seeing the error was to track every error message that gets printed.

Reply:
There is a way to capture Logger messages during tests, but I need to do some more research around it.  We can revisit these tests when we have a clear approach.